query.py

```diff
@@ -0,0 +1,188 @@
+import string
+
+import nltk
+
+stemmer = nltk.PorterStemmer()
+translator = str.maketrans('', '', string.punctuation)
+
+# TODO: Add NOT functionality
+# TODO: Optimise queries (skip pointers, size of postings, AND NOT)
+# TODO: Create test data set to check correctness
+class Query:
+    def __init__(self):
+        self.is_flipped = False
+        pass
+
+    def toDNF(self):
+        return Query()
+
+    def evaluate(self, inverted_index):
+        """
+        Return a list of documents that satisfies the query
+        """
+        raise NotImplementedError("evaluate not implemented")
+
+    def __str__(self):
+        return "Query"
+
+
+class QueryTerm(Query):
+    def __init__(self, term=""):
+        super().__init__()
+        self.term = stemmer.stem(term.strip().translate(translator).lower())
+
+    def evaluate(self, inverted_index):
+        # TODO maybe make this an iterator
+        _, docs = inverted_index.GetPostingListForTerm(self.term)
+        return set(docs)
+
+    def __str__(self):
+        return self.term
+
+
+class QueryOr(Query):
+    def __init__(self, ops):
+        super().__init__()
+        self.ops = ops
+
+    def evaluate(self, inverted_index):
+        union = self.ops[0].evaluate(inverted_index)
+        # if not self.operand1.is_flipped and not self.operand2.is_flipped:
+        for op in self.ops:
+            union.update(op.evaluate(inverted_index))
+        # compute intersection
+        return union
+
+    def __str__(self):
+        return "∨".join([op.__str__() for op in self.ops])
+
+
+class QueryAnd(Query):
+    def __init__(self, ops):
+        super().__init__()
+        self.ops = ops
+
+    def evaluate(self, inverted_index):
+        if len(self.ops) == 0:
+            return set()
+        # TODO: sort ops by size, evaluate from small to large
+        sets = [set(op.evaluate(inverted_index)) for op in self.ops]
+        intersection = sets[0]
+        for i in sets:
+            intersection.intersection_update(i)
+        return intersection
+
+    def __str__(self):
+        return "∧".join([op.__str__() for op in self.ops])
+
+
+class QueryNot(Query):
+    def __init__(self, op1:Query):
+        super().__init__()
+        self.operand1 = op1
+        self.is_flipped = not op1.is_flipped
+        # new_op = op1.copy()
+        # new_op.is_flipped = not new_op.is_flipped
+        # return new_op
+
+    def evaluate(self, inverted_index):
+        return self.operand1.evaluate(inverted_index)
+
+    def __str__(self):
+        return "¬{}".format(self.operand1)
+
+
+class Token:
+    AND = 1
+    OR = 2
+    NOT = 3
+    LB = 4
+    RB = 5
+
+class QueryParser:
+    @classmethod
+    def tokenize(cls, query_string: str):
+        tokens = []
+        for chunk in query_string.split(" "):
+            chunk = chunk.strip()
+            if chunk == "":
+                continue
+            elif chunk == "AND":
```

Comment:
Minor comment: 

In the assignment website, it states this:

"The operators will always be in UPPER CASE (lower case "and"s, "or"s and "not"s simply won't occur in your data (but you should probably bulletproof your code anyways)"

Maybe can add == "and", "or"...

Reply:
i prefer keeping the current upper case only operators, since it lets us use "and", "or", "not" as query terms